### PR TITLE
Vanity URLs with AWS result in bucket NOT_FOUND on  #2411

### PR DESCRIPTION
**Pull request by prb112**
Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java [hunk 1]
```diff
@@ -171,10 +171,16 @@ private static AmazonS3 getClient(boolean iam, String cosApiKeyProperty, String
      * @return
      */
     public boolean exists() {
-        boolean ex = client != null && client.doesBucketExistV2(bucketName);
-        if (ex == false) {
+        boolean ex = client != null;
+
+        // We only want to log a warning here, and assume it's true if the client exists
+        // in certain circumstances, a direct url to the bucket can be used. https://mybucketdemo123.s3.us.east-2.amazonaws.com
```
Comment (lmsurpre) [inline fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:177]:
```suggestion
        // We only want to log a warning here, and assume it's true if the client exists.
        // In certain circumstances, a direct url to the bucket can be used. https://mybucketdemo123.s3.us.east-2.amazonaws.com
```

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java [hunk 2]
```diff
@@ -171,10 +171,16 @@ private static AmazonS3 getClient(boolean iam, String cosApiKeyProperty, String
      * @return
      */
     public boolean exists() {
-        boolean ex = client != null && client.doesBucketExistV2(bucketName);
-        if (ex == false) {
+        boolean ex = client != null;
+
+        // We only want to log a warning here, and assume it's true if the client exists
```
Comment (lmsurpre) [inline fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:176]: So client will always be null if the bucket doesn't exist?

Comment (lmsurpre) [reply fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:176]: Or, if not, can you help me understand how will the end-user-facing error will change from this commit (in the case of a non-existent bucket)?

Comment (prb112) [reply fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java]: previously we were checking if the client was valid, and then doing the check.  In this case, we're only warning if the client doesn't exist.

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java [hunk 3]
```diff
@@ -171,10 +171,16 @@ private static AmazonS3 getClient(boolean iam, String cosApiKeyProperty, String
...
+        // in certain circumstances, a direct url to the bucket can be used. https://mybucketdemo123.s3.us.east-2.amazonaws.com
+        // versus an API enabled url e.g. https://s3.us.east-2.amazonaws.com
+        // These end up with TWO different responses, the former is false, and the latter is true.
+        if (!client.doesBucketExistV2(bucketName)) {
```
Comment (prb112) [inline fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:180]:
```suggestion
        if (ex || !client.doesBucketExistV2(bucketName)) {
```

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java (hunk 2 above)
Comment (prb112) [reply fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java]: I fixed line 180 to include ex, it was suppose to be !ex

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java [hunk 4]
```diff
@@ -171,10 +171,16 @@ private static AmazonS3 getClient(boolean iam, String cosApiKeyProperty, String
...
     public boolean exists() {
-        boolean ex = client != null && client.doesBucketExistV2(bucketName);
-        if (ex == false) {
+        boolean ex = client != null;
+
+        // We only want to log a warning here, and assume it's true if the client exists.
+        // In certain circumstances, a direct url to the bucket can be used. https://mybucketdemo123.s3.us.east-2.amazonaws.com
+        // versus an API enabled url e.g. https://s3.us.east-2.amazonaws.com
+        // These end up with TWO different responses, the former is false, and the latter is true.
+        if (ex || !client.doesBucketExistV2(bucketName)) {
```
Comment (prb112) [inline fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:180]:
```suggestion
        if (!ex || !client.doesBucketExistV2(bucketName)) {
```

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java]: I'm sorry, I'm still a bit confused.  The logic seems the same as the original logic (outside this PR) now.
I think maybe you were purposefully no longer relying on `client.doesBucketExistV2(bucketName)` and I was trying to understand why (and how that would bubble up in the errors we generate).

Comment (lmsurpre) [reply fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java]: ok, i had another look and i understand how its different now.

**Review by lmsurpre — APPROVED** [review]
Paul informed me that the system will still generate a reasonable error message for cases where the bucket doesn't actually exist.

However, I'd still like to better understand why the S3 SDK doesn't work with the bucket "vanity url" when the other methods do.  I wonder if its worth opening on stack overflow (or similar)...

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java [hunk 5]
```diff
@@ -171,10 +171,16 @@ private static AmazonS3 getClient(boolean iam, String cosApiKeyProperty, String
      * @return
      */
     public boolean exists() {
```
Comment (lmsurpre) [inline fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:173]: I wonder if we should update the javadoc to explain that it doesn't actually return false when the bucket doesn't exist now...

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java [hunk 6]
```diff
@@ -171,10 +171,16 @@ private static AmazonS3 getClient(boolean iam, String cosApiKeyProperty, String
      * @return
```
Comment (prb112) [inline fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:171]:
```suggestion
     * @return
     * @implNote to maintain the exists method behavior, we're only checking the client exists or does not  (true  or false), and warning if the bucket is not found as the S3 client does not do an exists properly when using a doesBucketExistV2 when using a vanity url to access the service on AWS.
```

File: fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java (hunk 5 above)
Comment (prb112) [reply fhir-bulkdata-webapp/src/main/java/com/ibm/fhir/bulkdata/provider/impl/S3Provider.java:174]: Updated

**Pull request merged**